### PR TITLE
[WIP]merged example into main project, not building

### PR DESCRIPTION
**Pull request by rishflab**
Add oauth functionality with r2d2 and diesel. 
Uses diesel ORM instead of raw sql queries.

### CODE REVIEW

File: src/bin/auth.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
+////#[macro_use]
```
Comment (SamuelMarks) [inline src/bin/auth.rs]: Remove this file rather than comment it out?

File: src/bin/r2d2_auth.rs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+#[derive(Debug, Serialize, Deserialize)]
+struct GetTokenReq {
+    name: String,
+    password: String,
+}
+
+const MAX_SIZE: usize = 262_144; // max payload size is 256k
+
+
+fn create_user((item, state): (Json<MyUser>, State<state::AppState>)) -> impl Future<Item = HttpResponse, Error = Error> {
```
Comment (SamuelMarks) [inline src/bin/r2d2_auth.rs]: Would be good to separate user and token creation to separate modules

File: src/config.rs [hunk 3]
```diff
@@ -1,11 +1,16 @@
...
+//    PgConnection::establish(database_url).expect(&format!("Error connecting to {}", database_url))
+//}
 
-pub fn establish_connection() -> PgConnection {
-    dotenv().ok();
-
-    let database_url = env::var("DATABASE_URL").expect("DATABASE_URL must be set");
-    PgConnection::establish(&database_url).expect(&format!("Error connecting to {}", database_url))
-}
+pub const SECRET: &str = "secret";
```
Comment (SamuelMarks) [inline src/config.rs]: Make this an environment variable, and unwrap it without a default specified for added security

File: src/db.rs [hunk 4]
```diff
@@ -0,0 +1,125 @@
+use actix_web::*;
```
Comment (SamuelMarks) [inline src/db.rs]: Separate user and token handling to their own distinct modules

File: src/schema.rs [hunk 5]
```diff
@@ -1,10 +1,7 @@
 table! {
-    user_tbl (email) {
-        title -> Nullable<Text>,
-        email -> Text,
-        password -> Nullable<Text>,
-        roles -> Nullable<Text>,
-        createdAt -> Nullable<Timestamptz>,
-        updatedAt -> Nullable<Timestamptz>,
+    users (id) {
```
Comment (SamuelMarks) [inline src/schema.rs]: Why did you add an `id` and remove the other attributes I had?

**Review by SamuelMarks — CHANGES_REQUESTED** [review]

**Comment by SamuelMarks** [comment]
@rishflab It's been 20 days since this PR opened, what's going on?

Just tried to build the `rishflab:oauth-r2d2` branch, and it doesn't build.

**Comment by rishflab** [comment]
> @rishflab It's been 20 days since this PR opened, what's going on?
> 
> Just tried to build the `rishflab:oauth-r2d2` branch, and it doesn't build.

yeh I broke it with recent commits. I am currently working on the android scaffold to get it AST based codgen working with it. 
I will fix this pull request either later tonight or first thing tomorrow as soon as the android scaffold is working with codegen. This thing won't take long to fix.

File: src/user/schema.rs [hunk 6]
```diff
@@ -1,5 +1,5 @@
 table! {
-    users (id) {
+    users (name) {
         id -> Text,
```
Comment (SamuelMarks) [inline src/user/schema.rs:3]: Remove the `id`

File: src/auth/routes.rs [hunk 7]
```diff
@@ -1,88 +1,36 @@
...
-//    fn handle(&mut self, msg: GetToken, _: &mut Self::Context) -> Self::Result {
-//        use self::schema::users::dsl::*;
-//
-//
-//    }
-//}
+        .from_err()
+        .and_then(|res| match res {
+            Ok(user) => Ok(HttpResponse::Ok().json(user)),
+            Err(_) => Ok(HttpResponse::InternalServerError().into()),
```
Comment (SamuelMarks) [inline src/auth/routes.rs:32]: RFC6749 define specific error codes, here is some inspiration from what I did 5 years ago: https://github.com/ng-bottle-libraries/oauth2-rfc6749/blob/master/rfc6749/oauth2_errors.py

**Review by SamuelMarks — CHANGES_REQUESTED** [review]

**Pull request merged**